synapse/rest/client/v2_alpha/account.py

```diff
@@ -592,6 +592,52 @@ def on_GET(self, request):
         finish_request(request)
 
 
+class AddThreepidMsisdnSubmitTokenServlet(RestServlet):
+    """Handles 3PID validation token submission for adding a phone number to a user's
+    account
+    """
+
+    PATTERNS = client_patterns(
+        "/add_threepid/msisdn/submit_token$", releases=(), unstable=True
+    )
+
+    def __init__(self, hs):
+        """
+        Args:
+            hs (synapse.server.HomeServer): server
+        """
+        super().__init__()
+        self.config = hs.config
+        self.clock = hs.get_clock()
+        self.store = hs.get_datastore()
+        self.identity_handler = hs.get_handlers().identity_handler
+
+    @defer.inlineCallbacks
+    def on_POST(self, request):
+        if not self.config.account_threepid_delegate_msisdn:
+            raise SynapseError(
+                400,
+                "This homeserver is not validating phone numbers. Use an identity server "
+                "instead.",
+            )
+
+        body = parse_json_object_from_request(request)
+        assert_params_in_dict(body, ["client_secret", "sid", "token"])
+
+        try:
+            # Proxy submit_token request to msisdn threepid delegate
+            response = yield self.identity_handler.proxy_msisdn_submit_token(
+                self.config.account_threepid_delegate_msisdn,
+                body["client_secret"],
+                body["sid"],
+                body["token"],
+            )
+            return 200, response
+        except HttpResponseException as e:
```

Comment:
this should be in the handler layer (not least because the [docstring](https://github.com/matrix-org/synapse/pull/6078/files#diff-09aa21e5bb05afad9baeb8141a43480fR464) says it is